[그리디] 이창희 로또 미션 3,4,5 단계 제출합니다

chxghee opened the pull request:
여우님 안녕하세요~~! 

저번 주신 리뷰 잘 받았습니다!!  정말 많은 도움이 되었어요 ㅎㅎ 
감사해요

이번에는 요구사항이 많이 추가 되면서 고민거리가 많아 졌던것 같아요 😅

그 중에 이번 미션을 진행하면서 가장 고민했던 점은

_Lottos 클래스 안에 수동로또필드와 자동로또필드로 구분하여 로또값들을 저장하는 것이 더 나은 선택일까?_
였습니다

제 나름대로 내린 결론은 

추후 반자동 로또 같은 요구사항이 추가된다면 필드를 또 만들어야 하고 필드를 추가하게 되면 Lottos를 사용하는 외부의 많은 로직변경이 필요하지만, 단일 로또 리스트로 관리하게 되면 생성로직만 추가해 주면 되고, 외부 의존하는 로직의 변경점이 비교적 적은 것 같다라는 결론을 내렸어요

그래서
Lottos 클래스에 필드를 추가하지 않고 수동 로또 일때 객체를 생성하는 정팩메만 추가하고  컨트롤러에서 manualLottos와 autoLottos를 따로 갖고 있도록 코드를 작성했습니다

그렇지만 지금은 Lottos라는 클래스를 삭제할까도 고민이 되는데요, 
왜냐하면 Lottos 라는 클래스는 자동 로또의 경우 자동 로또 구매 수량만큼 로또를 생성하기는 하지만 
수동 로또의 경우 그저 lotto 리스트를 wrapping만 해주고 있고, 그 외에는 별로 하는 일이 없기 때문이에요

하지만 일단은 일급 컬렉션을 쓰라는 프로그래밍 요구사항이 있어 Lottos를 유지해 놓긴 했어요.. ㅎㅎ

또 고민했던 점은 _비어있는 객체를 사용하는 것_이었는데요, (리드미의 5단계 리팩토링 부분에 적어 놓았어요)

현재 제 코드는 수동구매 개수가 0개 이거나 전부 수동구매를 하면, autoLottos, manualLottos 중 하나는 현재 리스트가 비어있는 객체가 생성되게 되는데 이때
optional 도입을 할까 생각했지만 optional을 도입하면 오히려 코드가 복잡해지고 추후 결과 집계 로직도 수정되어야 하는 상황이 발생해서
좀 더 직관적으로 처리할 수 있는 빈 객체를 생성하는 방식을 유지하기로 결정했어요


이렇게 제가 고민했던 부분에 대한 여우님의 생각이 궁금합니다!

다른 미흡한 점도 조언 부탁드려요~! 🙋‍♂️

chxghee committed:
Feat: 1단계 구현

chxghee committed:
Feat: 2단계 구현

chxghee committed:
Refactor: WinningNumbers 클래스 삭제

WinningNumbers 삭제 후 Lotto 클래스로 통합
그에 따른 의존 클래스 수정 및 테스트 수정

chxghee committed:
Refactor: 로또 번호 자료구조 TreeSet으로 변경
Lotto 클래스의 로또 번호를 담는 자료구조 변경 List -> TreeSet

chxghee committed:
Refactor: Lotto필드 수정 및 입력 오류 처리 컨트롤러로 이동
Lotto 의 필드 Set -> TreeSet으로 변경
InputHandler 삭제 후 컨트롤러에서 입력받고 오류 처리를 하도록 변경

chxghee committed:
Docs: 3단계 리드미 작성

chxghee committed:
Feat: 3단계 구현
보너스 볼 기능 추가

chxghee committed:
Test: 보너스 볼 관련 테스트 추가

chxghee committed:
Refactor: 정적팩토리메서드 이름 수정
컨벤션에 맞게 수정
Lotto.create() -> Lotto.from()
Lottos.createLottos() -> Lottos.of()

chxghee committed:
Docs: 4단계 리드미 작성

chxghee committed:
Feat: 4단계 구현
수동 로또 번호 생성 기능 추가

chxghee committed:
Refactor: 구매 금액 처리 관련 클래스 생성
기존 PurchaseAmount에서 관리한 구매금액 예외처리를 PurchasePrice로 분리하여 처리

chxghee committed:
Docs: 5단계 리팩토링 리드미 작성

chxghee committed:
Refactor: 돈 관련 필드 long타입으로 변경

BackFoxx commented on src/main/java/model/PurchaseAmount.java:
```diff
@@ -3,21 +3,44 @@
 public class PurchaseAmount {
 
     private static final int LOTTO_PRICE = 1000;
```
동일한 대상을 가리키는 LOTTO_PRICE 상수가 PurchaseAmount, DrawResults, PurchasePrice 세 군데에 중복으로 정의되어 있네요! 딱 한 군데에서 정의된 상수를 모두가 공통으로 쓸 수 있다면 관리가 더 효율적일 것 같습니다.

chxghee replied:
공통 상수들을 모아 놓은 클래스를 하나 만들어야 겠네요!

BackFoxx commented on src/main/java/model/Lottos.java:
```diff
@@ -12,14 +12,18 @@ private Lottos(List<Lotto> lottos) {
+            lottoList.add(Lotto.from(numbersGenerator.generate()));
         }
         return new Lottos(lottoList);
     }
```
수동 로또와 자동 로또를 어디서 어떻게 관리할지에 대한 고민을 많이 하셨군요 :)
로또라는 **추상적인** 개념 아래에 **구체적인** 분류로 수동 로또와 자동로또가 분류되어 있다는 점을 상기해보세요.
객체지향의 어떤 특성을 이용하면 List<Lotto> 하나에서 두 종류의 로또를 다 포함시킬 수 있을 것 같습니다.

chxghee replied:
위에서 주셨던 예시에서 뜨거운 물과 차가운 물 둘다 공통으로 물이라는 점에 주목을 하면,
수동로또와 자동로또는 둘 다 로또라는 공통점이 있네요   

그렇다면  수동로또와 자동로또를 **추상화해서 로또라는 클래스를 상속받는 형태로 만든다면** lottos 라는 클래스에서 다형성을 활용하여 하나의 리스트에서 두 종류의 로또를 포함할 수 있겠군요 !

예시로 설명을 해 주시니 이해가 잘 된것 같아요 감사합니다 ㅎㅎ

이번에는 이러한 방향으로 리팩토링을 해볼게요

BackFoxx replied:
멋지게 리팩토링 해주셨어요!
자동 로또와 수동 로또 객체가 분리되면서, 로또 생성 과정에 필요한 검증 과정이 AutoLotto에서 사라졌는데,
자동 로또는 시스템에서 자동으로 만들어주기 때문에 필요없다 판단하셔서 빼신 것 같아요.

```java
public class AutoLotto extends Lotto {

    private AutoLotto(TreeSet<LottoNumber> numbers) {
        super(numbers);
    }

    public static AutoLotto of(List<Integer> numbers) {
        List<LottoNumber> lottoNumbers = Utils.convertToLottoNumbers(numbers);
        return new AutoLotto(new TreeSet<>(lottoNumbers));
    }
}
```
하지만 어떤 동료 개발자가 로또 도메인을 잘 이해하지 못해서, 또는 실수로 자동 로또를 사용자가 만들 수 있게 잘못 작성할 수 있습니다. 그럼 사실상 99짜리 로또 번호도 만들 수 있어 다양한 문제가 발생할 것입니다.
이를 방지하려면 시스템이 만드는 자동 로또는 시스템 외에는 애초에 _접근_ 하지 못하게 적절한 제한을 걸어주어야 합니다. 어떻게 하면 좋을까요?

chxghee replied:
같은 모델 패키지에서만 생성 할 수 있게  public을 지우고 default로 접근 제한을 두면 괜찮을까요?

아니면 외부에서 번호들을 리스트로 받아오지 말고, 아예 내부적으로 랜덤한 6자리 번호를 생성하도록 하는 방법도 있겠군요!

어떤 방법이 더 좋은 방법인가요?? (다른 좋은 방법이 있다면 조언 주시면 감사하겠습니다!)

BackFoxx replied:
랜덤번호 생성이 AutoLotto 안으로 들어가면 자동차 경주의 숫자 생성과 비슷한 맥락의 고민이 또 생길 수 있습니다.
접근자 제한이 가장 편하고 효과적이겠네요! 코멘트를 남긴 의도도 접근자 제한이였습니다. 잘 캐치해주셔서 감사합니다 :)

BackFoxx commented on src/main/java/model/PurchaseAmount.java:
```diff
@@ -3,21 +3,44 @@
-        if (result < 1) {
-            throw new IllegalArgumentException("복권 구매의 최소 금액은 " + LOTTO_PRICE + "원 입니다!");
+    public static PurchaseAmount of(long purchasePrice, int manualPurchaseAmount) {
+        int purchaseAmount = (int) (purchasePrice / LOTTO_PRICE);
```
비즈니스 코드를 작성할 때, 크기가 큰 자료형에서 작은 자료형으로 형변환하는 것은 위험한 행동입니다.

말장난같은 가정이지만 현재 애플리케이션에서는 3조원(3,000,000,000,000)어치 로또를 구매하면 로또 구매 개수가 -1294967296로 오버플로가 발생합니다.

실제 비즈니스에서는 이런 캐스팅으로 인한 값 손실로 인해 많은 버그가 발생하고,
이런 오버플로는 예외가 발생하지도 않기 때문에 발견도 늦으며 원인 추적도 어렵습니다.

안전한 서비스를 만들기 위해서는 언제나 클라이언트가 일으킬 수 있는 최악의 상황을 가정하고, 이러한 상황이 발생하지 않도록 검증하는 작업이 필요합니다. :D

chxghee replied:
이러한 상황에서는 오버플로우가 날 수 있는 사용자의 입력이 발생하면(Integer.MAX_VALUE 보다 큰 값을 가지게 되면) 예외가 터지도록 해야겠군요
 
형 변환시 오버플로우는 항상 조심해야 겠네요 감사합니다!

BackFoxx commented on src/main/java/model/BonusBall.java:
```diff
@@ -0,0 +1,21 @@
+
+public class BonusBall {
+
+    private final int bonusNumber;
```
보너스 볼도 로또 번호인 만큼 1~45 사이의 수를 가져야 할텐데, 현재는 99짜리 보너스 볼도 입력할 수 있습니다.

하지만 그렇다고 BonusBall 생성자에 검증 로직을 넣자니 똑같은 검증 로직이 서로 다른 곳에 중복으로 구현되는 것이 우려될 것 같습니다.

다른 로또번호와 동일한 특징과 검증 로직을 가지면서도 BonusBall의 정체성을 유지하려면 어떻게 하면 좋을꺼요?

chxghee replied:
이 부분은 두가지의 방향이 떠오르는데요,

앞서 조언해 주셨던 방법처럼 로또숫자 하나를 의미하는 일급 컬렉션을 만들고 그 안에서 숫자 범위에 대한 검증로직을 구현하고, 
이것을 BonusBall에서 상속받도록 해서 보너스 볼 만의 검증 로직을 담는 방법과

저번 PR에 수동로또와 자동로또를 두가지의 팩토리 메서드로 생성하는 방식처럼 (정수기를 두개 만드는 것처럼)
로또숫자 하나를 의미하는 일급 컬렉션을 만들고 그 안에서 일반 로또 번호와 / 보너스 번호를 생성하는 방법이 떠올랐습니다

이 상황에서는 보너스 번호라는 것이 컨트롤러에서 6자리의 로또 번호와 별개로 관리가 되기 때문에
후자의 방법을 쓰는 것이 더 간단하고 더 이해하기 더 좋을 것 같다 라는 생각이 드네요

이러한 방향 외에 다른 좋은 방법이 있다면 조언 부탁드립니다!

BackFoxx replied:
일급 컬렉션은 이름에서 알 수 있듯 컬렉션(Collection)을 감싼 객체를 의미해요! 그래서 Lottos처럼 복수의 자료들을 담은 List나 Set 등을 불변하는 도메인 객체로 래핑했을 때 일급 컬렉션이라 부릅니다.
Lotto처럼 단일한 값을 래핑한 객체는 보통 값 객체(Value Object) 라고 불러요.

---

둘 다 좋은 방법이라고 생각합니다! 저는 전자와 후자의 방법을 합쳐서
LottoNumber를 BonusNumber가 상속하게 하여 LottoNumber의 공통점을 공유하되 고유한 정체성도 지키게 할 것 같아요.
그러나 BonusNumber의 생성을 LottoNumber를 통해서만 할 수 있게 해서, LottoNumber.bonusNumber()를 호출하면 그 결과로 BonusNumber를 반환하게 하겠습니다. 
이렇게 하면 LottoNumber만 사용해도 로또 번호와 보너스 번호를 다 만들 수 있다는 장점, 그리고 보너스 번호의 정체성을 분리할 수 있다는 장점을 다 누릴 수 있을 것 같습니다.

뭐가 더 좋은지는 저도 몰라요! :0 
두 가지 방법을 떠올리고, 각각의 장단점을 비교해서 더 나은 선택을 할 수 있다는 게 제일 중요한 것 같아요.

BackFoxx reviewed (changes requested):
안녕하세요! 이번 로또의 추가 요구사항이 꽤 복잡해서, 저도 처음 보고 :0 표정으로 보았습니다.

창희님께서 PR에 작성해주신 대로, 이번 추가 요구사항의 주요 쟁점은 수동 로또와 자동 로또입니다.
이 두 종류의 로또를 누가, 어떻게 정의하고 관리할 것인지 결정하는 전략이 정말 다양하기 때문에 많은 고민을 할 수 있는 지점이라고 생각합니다.

* 창희님께서 처음 구현하셨던, **lottos 안에 auto와 manual 두 개의 List<Lotto>를 두고 관리**하는 방식은 좋은 접근입니다.
이 방식을 실생활로 비유하면 하나의 정수기에 뜨거운 물이 나오는 출수구와 차가운 물이 나오는 출수구 두 개가 있는 것과 비슷합니다.
![image2](https://github.com/user-attachments/assets/0da9250a-2659-417c-aaf0-754fcd22dadb)
이 형태의 정수기는 다양한 온도의 물을 필요로 하는 사람들이 하나의 정수기만 찾아가도 된다는 장점이 있습니다.
그러나 미지근한 물처럼 새로운 온도의 물이 필요할 땐 출수구를 없애거나 추가하기 번거롭다는 단점이 있습니다.

* 창희님께서 PR에 제출해주신 최종적인 형태는 **auto와 manual용 Lottos를 따로 두고, 컨트롤러에서 관리하는 방식**입니다. 이 방식 또한 편리하고 좋은 방식입니다.
정수기로 비유하면, 차가운 물용 정수기, 뜨거운 물용 정수기가 따로 있는 것과 같습니다.
![image](https://github.com/user-attachments/assets/8ee11897-556c-4c77-9d77-c88ad1513638)
각 정수기가 명확히 한 종류의 물만 관리하기 때문에, 새로운 종류의 물이 필요해도 기존 정수기를 공사할 필요 없이 새로운 정수기를 사고 버리면 됩니다.
대신 정수기를 사용하는 사람 입장에서는, 자신이 필요로 하는 물의 종류에 따라 다른 정수기를 직접 찾아가야 하기 때문에 비교적 번거롭다는 단점이 있습니다.
현재 제출된 코드에서도 비슷한 맥락의 단점이 있습니다. LottoController가 auto용 Lottos와 manual용 Lottos를 각각 생성하고 사용함으로써, '로또들을 관리한다'라는 Lottos 객체의 역할을 LottoController가 다소 나눠가진 형태가 되어 있습니다.

상술한 두 형태의 구조는 각자의 장점과 단점을 서로가 반대로 가지고 있습니다.
두 구조의 장점만을 합한 멋진 구조가 있진 않을지 고민해보면 좋을 것 같아요.
이런 맥락에서 저는 새로운 형태의 접근을 하나 제안드리고 싶습니다.

바로 이런 모양의 정수기를 만들어 보는거에요.
![953c995763aec](https://github.com/user-attachments/assets/0c280543-b769-452e-a9d9-9ad69efd795f)
뜨거운 물, 차가운 물 둘 다 '물' 이라는 점을 활용하면, 꼭 출수구를 따로 두지 않고 하나의 파이프라인만으로도 두 종류의 물을 다 공급할 수 있을 것 같습니다.
또 새로운 온도의 물이 추가되더라도, 그 대상이 '물'이기만 하면 똑같은 출수구로 공급해줄 수 있을 거에요.
객체지향의 이점을 살려서 이 정수기같은 Lottos 객체를 한 번 만들어 보아요.
만드는 과정에서 또 다양한 고민과 의문이 생기실 수 있는데, 그 땐 또 이야기를 나누어 보아요. :)

chxghee committed:
Refactor: Lotto 클래스 추상화
AutoLotto, ManualLotto가 상속받도록 변경
test코드 변경

chxghee committed:
Refactor: 로또번호 일급 컬렉션 추가
BonusBall을 클래스를 삭제하고 LottoNumber 일급 컬렉션을 추가하여 로또 번호 범위에 대한 유효성 검사를 한곳에서 진행하도록 수정

chxghee committed:
Test: 테스트 코드 수정 및 추가

chxghee commented:
일단 주신 피드백 토대로 리팩토링 해보았습니다! 

그림까지 넣어 주시면서 설명을 해 주셔서 더욱 이해가 쉬웠던것 같네요 감사합니다!! 

우선 Lotto 클래스를 추상화 해서 온수와 냉수를 한 출수구에서 공급하듯, 수동 로또와 자동 로또를 Lottos 클래스에서 하나의 리스트로 관리할 수 있게 바꾸었어요

사실 프로그래밍을 공부 하면서 추상화를 해서 클래스를 상속한다는 것이 좋은건 알겠는데 귀찮기도 하고 언제 사용해야 하나 하는 기준이 불분명 해서 잘 사용하지는 않았었는데요,

이번에 확실히 이전 코드와 비교를 하여 필요성을 깨닫게 되고 도입을 하는 과정을 거쳐 보니까 언제 써먹을 수 있고 이래서 중요하구나 하는걸 몸소 체감했던 시간이 된 것 같아요

이번 미션에는 모든 요구사항에 대해 일급 컬렉션을 써서 뭔가 클래스 들을 엄청 잘게잘게 나뉜 느낌이라 구조가 간단하다고는 생각이 안드는데요,
그래도 이전보다 더욱 객체지향적인 코드가 된 것 같은 느낌이 듭니다! 

제가 이번에 제출한 코드에 더욱 개선할 점이 있다면 조언해 주신다면 감사하겠습니다!
이번 미션 리뷰해 주셔서 감사합니다! 🙇‍♂️

chxghee committed:
Refactor: 필요없는 메서드 및 필드 삭제

BackFoxx commented on src/main/java/utils/LottoConstants.java:
```diff
@@ -0,0 +1,9 @@
+package utils;
+
+public final class LottoConstants {
```
여러 곳에서 사용하는 상수를 묶어주신 시도는 좋았어요!
그런데 LottoConstants라는 클래스가 따로 존재하도록 구현하신 배경이 궁금해요. 로또 도메인 클래스 안에 있는 것과 비교해 어떤 장단점이 있을까요?

chxghee replied:
로또 도메인 클래스에 모든 상수를 정의해 놓고 사용한다면 
의미적으로 이해도 쉽고 응집도도 높은 것이 장점이지만,
다른 도메인 클래스에서 이 상수를 사용한다면, 로또 클래스에 의존하게 되는 것이 단점이라고 생각합니다

그래서 다른 도메인 클래스에서도 많이 사용되는 상수를 
다른 로직이 없고 상수만을 취급하는 외부 클래스에서 정의해 놓는다면 관리도 쉽고 재사용하기 좋은것 같아서 외부 클래스에 선언하는 것으로 코드를 리팩토링 해 보았어요

혹시 다른 장단점이 있을까요?

BackFoxx replied:
말씀해주신 내용이 맞습니다! 특정 도메인 객체에서 그 객체와 관련된 상수를 관리한다면 관심사를 한 곳에 모을 수 있다는 장점이 있느나, 외부에서 그 객체에 의존할 일도 많아집니다.
그래서 다양한 곳에서 사용할 상수들은 어떤 도메인에도 속하지 않는 곳에서 관리하는 것이 일반적이고, 스프링과 같은 프레임워크를 활용한다면 아예 코드레벨에서 분리하여 yml같은 별도의 설정파일로 상수를 관리하기도 해요.

BackFoxx commented on src/main/java/model/LottoNumber.java:
```diff
@@ -0,0 +1,59 @@
+
+    private static void validateLottoNumberBound(int number) {
+        if (number > LOTTO_MAX_NUMBER || number < LOTTO_MIN_NUMBER) {
+            throw new IllegalArgumentException("로또 번호는 " + LOTTO_MIN_NUMBER + "이상 " + LOTTO_MAX_NUMBER + "이하의 정수입니다!");
```
보통 이렇게 문자열을 연결하는 작업을 할 때 '+' 연산자를 쓰지 말라고 합니다. 그 이유를 함께 알아보면 자바를 깊게 이해하는 데에 도움이 될 것 같습니다. 아래 질문들에 답해주세요. :)

* 사람들이 이야기하는 + 연산의 문제점은 무엇일까요?
* 그 문제점이 위 경우에도 적용될까요? 적용된다면/안된다면 그 이유는 무엇인가요?
* 문제점이 적용되는 경우는 언제이고, 대체제는 무엇인가요?

chxghee replied:
생각하지 못한 부분이었는데 이번에 이런 이슈들이 있구나 하는 걸 알게 되었네요!

일단 알아본 내용을 토대로 답변해 보겠습니다

#### 1. 사람들이 이야기하는 + 연산의 문제점은 무엇일까요?

> String 클래스는 변경 불가능한 클래스 이기 때문에 + 연산을 하게 되면 _내부적으로 `StringBuilder` 클래스가 생성된 후  문자열에 append하고 결과를 반환하는_ 과정을 거치게 됩니다. 
때문에 + 로 문자열을 합치게 되면 항상 새로운 문자열이 생성되기 때문에 _메모리의 낭비가 발생하게 됩니다._

#### 2. 그 문제점이 위 경우에도 적용될까요? 적용된다면/안된다면 그 이유는 무엇인가요?

> 위와 같은 경우에는 문제가 발생하지 않습니다.
왜냐하면 ....  
문자열 + 연산의 피연산자가 `"문자열"`과 `final 로 선언된 상수`일 경우는  컴파일 시점에 상수값이 확정되기 때문에 _컴파일러가 컴파일 시점에 문자열 리터럴로 합치게 됩니다._


#### 3. 문제점이 적용되는 경우는 언제이고, 대체제는 무엇인가요?
> 문제가 발생하는 상황은 런타임 시 문자열 + 연산이 여러번 반복 될때 문제가 발생합니다.

조금 더 자세하게 설명하자면,

런타임에 한 번만 실행되는 문자열 + 연산인 경우 딱 한번  `StringBuilder` 클래스가 생성되고 문자열을 합치기 때문에 큰 상관이 없지만,
런타임에 여러번 실행되는(반복이 되는) 문자열 + 연산의 경우 매번 `StringBuilder` 클래스가 생성되기 때문에 성능 문제가 발생합니다

이러한 상황에서 + 연산자를 대체하기 위해 반복되는 곳 바깥에서 `StringBuilder 클래스`를 생성하고 반복되는 로직 안에서 사용해 문자열을 합치게 된다면 `StringBuilder 클래스`를 한 번만 생성하기 때문에 메모리 문제를 해결할 수 있습니다.

BackFoxx replied:
남겨주신 내용이 맞습니다! 불변하는 상수값들과 문자열의 + 연산을 컴파일 시점에 미리 수행해 최적화하는 기법을 constant folding이라고 합니다.
그런데 이 말을 반대로 생각해보면, 문자열에 대한 + 연산들을 컴파일 시점에 몰아서 해야 한다는 것이므로 프로그램에 불필요한 부담을 주는 요인이 되기도 합니다.
창희님의 코드를 예시로 들어서 예외 메시지를 위처럼 + 연산으로 작성한 코드가 천만 군데 있다고 하면, 실제 예외 발생시에만 사용하고 평소에는 쓰지 않을 문자열들을 합쳐두기 위해 컴파일 시점에 천만 개의 연산을 해야 하는 것입니다.
물론 요즘 수준의 컴퓨터에서 문자열 연산 정도는 수억 번 수행해도 끄떡없지만, 이를 신경쓰면서 작성한 코드와 그렇지 않은 코드는 장기적으로 큰 차이가 납니다. :)

BackFoxx commented on src/main/java/model/LottoNumber.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    @Override
+    public boolean equals(Object o) {
```
number 값이 같으면 같은 LottoNumber로 인식할 수 있게 equals 메서드를 재정의 해주셨네요.
그런데 동등성 보장을 위해 equals만 재정의하면 예상치 못한 곳에서 낭패를 볼 수 있습니다. 그 이유가 무엇일까요?

chxghee replied:
hashCode()도 같이 재정의를 해야 하는군요!

HashSet과 HashMap 등을 사용할 때 
먼저 hashCode() 를 사용해서 버킷에 분류를 해 빠르게 값이 같은지 확인할 수 있게 합니다

하지만,
equals는 ture를 반환하는데, hashCode()가 False를 반환하게 되면 같은 값을 가진 객체일 지라도 false를  반환할 수 있어 기대와 다르게 동작할 수 있습니다

BackFoxx replied:
맞습니다. hashCode()를 오버라이딩하지 않으면 인스턴스의 메모리 주소를 해시 함수에 돌려 값을 만들기 때문에 hash를 사용해 동등성 연산을 하는 경우 사실상 == 연산이 됩니다. 면접 질문으로 자주 나오는 내용이니 숙지해두시면 좋습니다!

BackFoxx reviewed (changes requested):
상속을 사용해서 멋진 리팩토링을 해주셨네요 :) 리팩토링을 하고 나니 더 깔끔하고 정석적인 것 같은데, 오히려 번잡해진 것 같고 뭔가 이렇게까지 해야 하나 싶은 느낌이 드실 것도 같습니다.

사실 '상속'이라는 기술은 좀 번거롭습니다. 상속받는 객체가 상속 하는 객체에게 크게 의존하게 되는데, 객체가 다른 객체에게 강한 의존을 가지고 있을 때 가질 수 있는 문제점을 상속 또한 똑같이 가지고 있습니다. 그래서 변경이 자주 일어나는 실무환경에서는 아예 안티패턴으로 취급하기도 합니다.
그런데 '상속'을 사용해서 '프록시'라는 구조를 만들기 시작하면 생산성이 갑자기 폭발하며 칼춤을 추기 시작하는데, 상속을 주축으로 하는 다형성을 정말 극한의 극한까지 활용한 결과물이 바로 4월 말부터 학습하실 '스프링 프레임워크'입니다.
앞으로 시간이 지날수록 더 깊숙히 들어올 개념인 만큼, 이번 미션을 통해 상속과 미리 친해질 수 있는 시간을 드릴 수 있어서 더없이 기쁩니다. :D

객체지향적인 코드는 이미 잘 작성해주셔서, 신경쓰지 않으면 지나치기 쉬운 자바 개념 중심으로 마지막 코멘트를 남겼습니다. 코드 수정은 하지 않으셔도 되니 질문에 대한 답을 한 번씩 찾아봐주신 후 의견을 남겨주세요.

BackFoxx commented on src/main/java/model/Ranking.java:
```diff
@@ -4,25 +4,41 @@
 
 public enum Ranking {
```
Enum은 필드로 람다 함수를 두고 사용할 수 있습니다.
이를 활용하면 ranking 계산 코드를 더 간단히 표현할 수 있을 것 같아요.

chxghee replied:
오늘 스터디에서도 이 부분에 대한 이야기가 나왔는데,
이렇게 함수형 인터페이스를 사용하는 것은 아직 익숙하지 않아서 피하지 말고 더욱 연습을 해 보아야 할 것 같아요
리뷰 감사합니다 😄

BackFoxx reviewed (approved):
곧 시험준비도 해야하는데 많은 공부 하시느라 수고하셨어요! 앞으로도 재미있게 오래오래 공부해요 :)

chxghee commented:
바쁘신 와중 신경써서 리뷰해 주셔서 정말 감사해요~~

덕분에 많은 도움이 되었습니다! 🙇‍♂️
